### PR TITLE
update verilator coverage to work if vl_finish ends the testrun

### PR DESCRIPTION
**Pull request by forensicgarlic**
update inspired by https://github.com/cocotb/cocotb/issues/1478#issuecomment-903345076 However, through trial and error I discovered that sometimes tests end with vl_finish, and other times not, so both write("coverage.dat") lines are needed in verilator.cpp.

### CODE REVIEW

**Comment by marlonjames** [comment]
`vl_finish()` will stop the simulation immediately and exit if it is called twice.
Do any of your testbenches use the `$finish` system task? Because that calls `vl_finish()` in Verilator.

**Comment by marlonjames** [comment]
The better approach is probably to either replace the call to `vl_finish()` with `Verilated::gotFinish(true);`, or to register a Verilator flush or exit callback to ensure the coverage is written out before exit.

On second though, because we break out of the while loop anyway, we might not need to call either function, just break.

**Comment by forensicgarlic** [comment]
the tests i'm running seeing the need for writing before vl_finish is called, and outside that loop checking the time left are a mix of the things in the cocotb repo in examples/ and tests/ I'd have to go back and run them again to see which tests precisely were which. 

Happy to update for a different exit path. Do you prefer just removing the vl_finish() call?

**Comment by marlonjames** [comment]
Yes, I would just remove the `vl_finish()` call.

**Comment by forensicgarlic** [comment]
gave that a shot, and verified with test_1279 and adding `--coverage` to the Makefile.verilator `EXTRA_ARGS` variable, without the change, no coverage.dat was generated, but with the change it was.

File: documentation/source/newsfragments/1471.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+fix verilator.cpp for writing coverage files when calls to vl_finish end tests
```
Comment (marlonjames) [inline documentation/source/newsfragments/1471.bugfix.rst:1]:
```suggestion
Fixed Verilator not writing coverage files in some cases.
```

Also update this file name to `1478.bugfix.rst`

Comment (forensicgarlic) [reply documentation/source/newsfragments/1471.bugfix.rst]: fixed in recent push.

File: documentation/source/newsfragments/1478.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+fix verilator.cpp for writing coverage files when calls to vl_finish end tests
```
Comment (marlonjames) [inline documentation/source/newsfragments/1478.bugfix.rst:1]:
```suggestion
Fixed Verilator not writing coverage files in some cases.
```

File: documentation/source/newsfragments/1478.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+fix Verilator not writing coverage files in some cases
```
Comment (marlonjames) [inline documentation/source/newsfragments/1478.bugfix.rst:1]:
```suggestion
Fixed Verilator not writing coverage files in some cases
```
Apologies for nitpicking but these end up in the release notes.

Comment (forensicgarlic) [reply documentation/source/newsfragments/1478.bugfix.rst]: my apologies, i missed this on the first request.

File: documentation/source/newsfragments/1478.bugfix.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Fixed Verilator not writing coverage files in some cases
```
Comment (marlonjames) [inline documentation/source/newsfragments/1478.bugfix.rst:1]:
```suggestion
Fixed Verilator not writing coverage files in some cases.
```

**Review by marlonjames — APPROVED** [review]

**Pull request merged**